extra_data/caldata.py

```diff
@@ -0,0 +1,1016 @@
+    or optional. A required key is always present and accordingly must
+    have a default value. An optional key will be present if it is
+    specified in any of the layers or has a default value. If however
+    its value is set to None explicitly in any layer, the default value
+    is ignored.
+
+    A parameter name has two representations, the canonical database
+    form (e.g. Sensor Bias Voltage) and a key form, which must be a
+    valid name for a Python symbol. By default, the key form is the all
+    lower-case version of the database name with spaces replaced by
+    underscores (e.g. sensor_bias_voltage). The key form is expected in
+    the parameter layers.
+
+    Example:
+
+    .. code-block:: python
+
+       cond = OperatingCondition(
+           dict(key_a=1, key_b=2),
+           dict(key_a=3, key_c=None)
+       )
+
+       cond.set_required('Key A', -1)
+       cond.set_optional('Key B')
+       cond.set_optional('Key C', 5)
+
+       assert cond == {'Key A': 1, 'Key B': 2}
+    """
+
+    def __init__(self, *param_layers):
+        """Initialize a new operating condition.
+
+        Args:
+            *param_layers (list of dict): Parameters layers with values
+                with the lowest layer (passed first) taking precedence
+                for any given parameter, which must be referred to by
+                their key name.
+        """
+
+        self._param_layers = param_layers
+
+    def set_required(self, db_name, default, key_name=None):
+        """Set a required parameter.
+
+        Args:
+            db_name (str): Parameter name in database form.
+            default (Any): Default value for this parameter if not found
+                in any parameter layer.
+            key_name (str, optional): Parameter name in key form, if
+                omitted obtained from database form.
+
+        Returns:
+            None
+        """
+
+        assert default is not None, \
+            'Required condition parameter must have a default'
+        self.set_optional(db_name, default, key_name)
+
+    def set_optional(self, db_name, default=None, key_name=None):
+        """Set an optional parameter.
+
+        Args:
+            db_name (str): Parameter name in database form.
+            default (Any, optional): Default value for this parameter if
+                not found in any parameter layer. If omitted, the
+                parameter is not added at all in this case.
+            key_name (str, optional): Parameter name in key form, if
+                omitted obtained from database form.
+
+        Returns:
+            None
+        """
+
+        if key_name is None:
+            key_name = db_name.lower().replace(' ', '_')
+
+        for layer in self._param_layers:
+            if key_name in layer:
+                if layer[key_name] is not None:
+                    self[db_name] = layer[key_name]
+                return
```

Comment:
LGTM.com complains here about `This 'for' statement has a redundant 'else' as no 'break' is present in the body.`

Interesting it does not realize the `return` statement. A `break` would be identical in this case, but now I'm curious on an academic level: Do you think it should be generally preferred in such a case?

Reply:
I guess LGTM.com is right in that the `else` is equivalent to simply putting the code after the for loop. You could argue that using else makes the intent clearer, but I'm not sure I would argue that.

For this use case - do something side-effect-ish with one thing in this collection, or fall back to doing a default thing - I'd marginally lean towards a for/else with a break. I think that makes refactoring easier e.g. if you want to add another step afterwards which happens regardless of this step. But if the loop & the fallback were to retrieve a value, I'd lean towards `return` (and no `else`).